console-src/src/enhancers/withOAuthClients.tsx

```diff
@@ -0,0 +1,190 @@
+  CreateOAuthClientPayload,
+  UpdateOAuthClientPayload,
+  DeleteOAuthClientPayload
+} from "../types/oAuthClient";
+import {
+  fetchClientsQuery,
+  createClientMutation,
+  updateClientMutation,
+  deleteClientMutation
+} from "../queries";
+
+const { baseAPIUrl, clientId, redirectUri } =
+  process.env.NODE_ENV === "production" ? prodConfig : devConfig;
+
+const STATE_KEY = "state";
+const VERIFIER_KEY = "verifier";
+const CLIENT_SCOPE = "clients";
+const CODE_QUERY_PARAM = "code";
+
+sessionStorage.setItem(
+  STATE_KEY,
+  sessionStorage.getItem(STATE_KEY) || (Math.random() + "").substring(2)
+);
+sessionStorage.setItem(
+  VERIFIER_KEY,
+  sessionStorage.getItem(VERIFIER_KEY) || (Math.random() + "").substring(2)
+);
+
+const client = new Client({
+  baseUrl: baseAPIUrl,
+  clientId,
+  redirectUri,
+  scopes: [CLIENT_SCOPE],
+  state: sessionStorage.getItem(STATE_KEY) || "",
+  verifier: sessionStorage.getItem(VERIFIER_KEY) || ""
+});
+
+interface OAuthClientsContext {
+  isLoading: boolean;
+  oAuthClients: OAuthClient[];
+  createClient: (payload: CreateOAuthClientPayload) => Promise<Object> | void;
+  updateClient: (payload: UpdateOAuthClientPayload) => Promise<Object> | void;
+  deleteClient: (payload: DeleteOAuthClientPayload) => Promise<Object> | void;
+}
+
+interface State extends OAuthClientsContext {
+  isAuthenticated: boolean;
+}
+
+const { Provider, Consumer } = React.createContext<OAuthClientsContext>({
+  isLoading: false,
+  oAuthClients: [],
+  createClient: () => {},
+  updateClient: () => {},
+  deleteClient: () => {}
+});
+
+class OAuthClientsProvider extends Component<{}, State> {
+  fetchClients: () => Promise<any>;
+  createClient: (payload: CreateOAuthClientPayload) => Promise<any>;
+  updateClient: (payload: UpdateOAuthClientPayload) => Promise<any>;
+  deleteClient: (payload: DeleteOAuthClientPayload) => Promise<any>;
+
+  constructor(props: {}) {
+    super(props);
+
+    this.fetchClients = async () => {
+      this.setState({
+        isLoading: true
+      });
+
+      const { viewer } = await client.graphQL.rawQuery(fetchClientsQuery);
+
+      this.setState(
+        (state: State): State => ({
+          ...state,
+          isLoading: false,
+          // Currently rawQuery return type does not include clients
+          // @ts-ignore
+          oAuthClients: viewer.clients
+        })
+      );
+    };
+
+    this.createClient = async (payload: CreateOAuthClientPayload) =>
+      client.graphQL.rawQuery(createClientMutation, payload);
+
+    this.updateClient = async (payload: UpdateOAuthClientPayload) =>
+      client.graphQL.rawQuery(updateClientMutation, payload);
+
+    this.deleteClient = async (payload: DeleteOAuthClientPayload) =>
+      client.graphQL.rawQuery(deleteClientMutation, payload);
+
+    this.state = {
+      isAuthenticated: false,
+      isLoading: false,
+      createClient: this.createClient,
+      updateClient: this.updateClient,
+      deleteClient: this.deleteClient,
```

Comment:
Why are these methods stored in component's state?

Reply:
Same as above 🤷‍♂ 
https://reactjs.org/docs/context.html#updating-context-from-a-nested-component

It is also quite confusing to me but on the other hand I have always done it this way when using the most recent context api

Reply:
same as above, it seems that it is not necessary, so I have changed it here: a7bf1ea8d0780811180cdae92c60404325cc521c